### PR TITLE
Fix/csv export attachments

### PR DESCRIPTION
**Pull request by hjJunior**
### Fixed
- Imports notes of stories

### Added
- Imports/exports of stories 
  - Documents
  - Tasks

Related PR on https://github.com/Codeminer42/cm42-central/pull/406

### CODE REVIEW

File: spec/support/rails_app/.env [hunk 1]
```diff
@@ -0,0 +1 @@
+CLOUDINARY_URL=cloudinary://username:password@localhost
```
Comment (adbatista) [inline spec/support/rails_app/.env]: add new line at the end of file

File: spec/support/rails_app/Gemfile [hunk 2]
```diff
@@ -11,6 +11,9 @@ gem 'central-support', path: File.expand_path("../../../..", __FILE__)
 gem 'devise'
 gem 'pg'
 gem 'friendly_id'
+gem 'attachinary'
+gem 'cloudinary'
+gem 'dotenv-rails'
```
Comment (adbatista) [inline spec/support/rails_app/Gemfile]: this should go under development group

File: lib/central/support/concerns/story_concern/associations.rb [hunk 3]
```diff
@@ -9,6 +9,11 @@ module Associations
           belongs_to :owned_by, class_name: 'User'
 
           has_many :users, through: :project
+          has_many :tasks
+          has_attachments :documents,
+                          accept: %i[raw jpg png psd docx xlsx doc xls pdf],
+                          maximum: 10
```
[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/central/support/concerns/story_concern/associations.rb]: I just got this number from central, should I remove this limit?

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -8,6 +8,19 @@
...
 
+# Add additional requires below this line. Rails is not loaded until this point!
+require 'vcr'
+require 'webmock'
+
+VCR.configure do |config|
+  config.cassette_library_dir = 'fixtures/vcr_cassettes'
+  config.hook_into :webmock # or :fakeweb
+  config.ignore_localhost = true
+  config.allow_http_connections_when_no_cassette = true
```
[the remark this answers is not in the document]
Comment (hjJunior) [reply spec/rails_helper.rb]: I just added to reproduce the same env from [central](https://github.com/Codeminer42/cm42-central/blob/6a97cba8e84d23a717c8c79ad6f541444acda57e/spec/rails_helper.rb#L11)

File: lib/central/support/concerns/story_concern/associations.rb [hunk 5]
```diff
@@ -9,6 +9,11 @@ module Associations
           belongs_to :owned_by, class_name: 'User'
 
           has_many :users, through: :project
+          has_many :tasks
+          has_attachments :documents,
+                          accept: %i[raw jpg png psd docx xlsx doc xls pdf],
```
[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/central/support/concerns/story_concern/associations.rb]: can I add all of this?

Formato do documento | Extensão de arquivo
-- | --
Texto ODF | *.odt
Modelo de texto ODF | *.ott
Documento mestre ODF | *.odm
Documento HTML | HTML
Modelo de documento HTML | *.oth
Planilha ODF | *.ods
Modelo de planilha ODF | *.ots
Desenho ODF | *.odg
Modelo de desenho ODF | *.otg
Apresentação ODF | *.odp
Modelo de apresentação ODF | *.otp
Fórmula ODF | *.odf
Banco de dados ODF | *.odb
Extensão do LibreOffice | *.oxt

From [Libreoffice help](https://help.libreoffice.org/Common/XML_File_Formats/pt-BR)

[the remark this answers is not in the document]
Comment (adbatista) [reply lib/central/support/concerns/story_concern/associations.rb]: sure, and now I noticed that is a dummy project

can you send a PR to the original repo?

File: lib/central/support/concerns/story_concern/associations.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (adbatista) [reply lib/central/support/concerns/story_concern/associations.rb]: no need I Just wanna know why only 10 attachments

File: spec/rails_helper.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adbatista) [reply spec/rails_helper.rb]: you don't need this, even the original project should nott have this

File: lib/central/support/concerns/story_concern/associations.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/central/support/concerns/story_concern/associations.rb]: Of course, I'll send to open PR related

[the remark this answers is not in the document]
Comment (hjJunior) [reply lib/central/support/concerns/story_concern/associations.rb]: I'll send to related PR on original repo

File: lib/central/support/concerns/project_concern/csv.rb [hunk 6]
```diff
@@ -24,31 +23,39 @@ def from_csv(csv_string)
...
+                when 'Document'
+                  story.documents << ::Attachinary::File.new(JSON.parse(value.gsub '=>', ':'))
+                when 'Task'
+                  next_value = row[index+1].presence
+                  next if next_value.blank?
+                  tasks.unshift(Task.new(name: value, done: next_value == 'completed'))
+                end
               end
-              story.description = "#{story.description}\n\nTasks:\n\n#{tasks.join("\n")}" unless tasks.empty?
+              story.description = story.description
```
Comment (talyssonoc) [inline lib/central/support/concerns/project_concern/csv.rb]: Why is this needed?

File: lib/central/support/concerns/project_concern/csv.rb [hunk 7]
```diff
@@ -24,31 +23,39 @@ def from_csv(csv_string)
...
               story.project.suppress_notifications = true # otherwise the import will generate massive notifications!
+              story.tasks = tasks
+              story.notes = story.notes.from_csv_row(row)
```
Comment (talyssonoc) [inline lib/central/support/concerns/project_concern/csv.rb]: How does it work? It seems weird for me to get the return of the relation and save in the relation.

File: lib/central/support/concerns/project_concern/csv.rb [hunk 8]
```diff
@@ -24,31 +23,39 @@ def from_csv(csv_string)
...
 
-              # Generate notes for this story if any are present
-              story.notes.from_csv_row(row)
-
+              row_state = ( row_attrs["Current State"] || 'unstarted').downcase
+              if Story.available_states.include?(row_state.to_sym)
+                story.state = row_state
+              end
+              story.accepted_at = row_attrs["Accepted at"]
+              story.save
```
Comment (talyssonoc) [inline lib/central/support/concerns/project_concern/csv.rb]: What happens if this `save` fails?

File: spec/central/support/project_spec.rb [hunk 9]
```diff
@@ -119,6 +119,8 @@
 
   describe 'CSV import' do
     let(:project) { create :project }
+    let(:task) { double('task') }
+    let(:tasks) { [task, task] }
```
Comment (talyssonoc) [inline spec/central/support/project_spec.rb]: It seems odd to use the same double twice, can we create a `task_1` and `task_2`?

File: spec/support/rails_app/db/migrate/20181030162426_create_attachinary_tables.attachinary.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# This migration comes from attachinary (originally 20120612112526)
+class CreateAttachinaryTables < ActiveRecord::Migration
```
Comment (talyssonoc) [inline spec/support/rails_app/db/migrate/20181030162426_create_attachinary_tables.attachinary.rb]: Do we need it here?

File: lib/central/support/concerns/project_concern/csv.rb (hunk 8 above)
Comment (hjJunior) [reply lib/central/support/concerns/project_concern/csv.rb]: There is a rescue on start the async service, so if tails we got the error and pass as notification to UI

File: spec/support/rails_app/db/migrate/20181030162426_create_attachinary_tables.attachinary.rb (hunk 10 above)
Comment (hjJunior) [reply spec/support/rails_app/db/migrate/20181030162426_create_attachinary_tables.attachinary.rb]: Yep, because the relationship of documents
